src/qmapshack/gis/rte/router/CRouterOptimization.cpp

```diff
@@ -0,0 +1,322 @@
+/**********************************************************************************************
+    Copyright (C) 2019 Henri Hornburg hrnbg@t-online.de
+
+    This program is free software: you can redistribute it and/or modify
+    it under the terms of the GNU General Public License as published by
+    the Free Software Foundation, either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU General Public License for more details.
+
+    You should have received a copy of the GNU General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+**********************************************************************************************/
+
+#include "CRouterOptimization.h"
+#include <gis/rte/router/CRouterSetup.h>
+#include <GeoMath.h>
+#include <helpers/CProgressDialog.h>
+#include <QDebug>
+#include <QElapsedTimer>
+#include <QRandomGenerator>
+
+CRouterOptimization::CRouterOptimization()
+{
+    routerOptions = CRouterSetup::self().getOptions();
+}
+
+int CRouterOptimization::optimize(SGisLine &line)
+{
+    /********************************************
+    * DEBUG
+    * ******************************************/
+    QElapsedTimer timer;
+    QElapsedTimer totalTimer;
+    totalTimer.start();
+    qreal timeEveryIterationTest = 0;
+    qreal timePreprocessing = 0;
+    qreal timeRouter = 0;
+    qreal timeDisplay = 0;
+    int successfulTries = 0;
+    int totalTries = 0;
+    /********************************************
+    * ******************************************/
+
+    checkRouter();
+    if(!CRouterSetup::self().hasFastRouting())
+    {
+        return -1;
+    }
+    if(line.length() < 4)
+    {
+        return 0; //There is nothing to optimize
+    }
+
+    CProgressDialog progress(CProgressDialog::tr("Optimizing route"), 0, line.length(), nullptr);
+    progress.setWindowFlag(Qt::Tool, true);
```

Comment:
@kiozen, is there a way to keep the CProgressDialog in front? This doesn't do it.

Reply:
No as it is a non-modal dialog. (or was it modal?) Anyways CProgressDialog sucks big time.